File: flink/KafkaCDCPrintJob.py
from pyflink.common import WatermarkStrategy
from pyflink.common.serialization import SimpleStringSchema
from pyflink.datastream import StreamExecutionEnvironment
from pyflink.datastream.connectors.kafka import (
    KafkaSource,
    KafkaOffsetsInitializer,
)


def main():

    env = StreamExecutionEnvironment.get_execution_environment()

    source = (
        KafkaSource.builder()
        .set_bootstrap_servers("kafka:29092")
        .set_topics("sales.public.customers")
        .set_group_id("pyflink-cdc-print-test-v3")
        .set_starting_offsets(KafkaOffsetsInitializer.earliest())
        .set_value_only_deserializer(SimpleStringSchema())
        .build()
    )

    stream = env.from_source(
        source,
        WatermarkStrategy.no_watermarks(),
        "Kafka CDC Source",
    )

    stream.print()

    env.execute("PyFlink Kafka CDC Print Test")


if __name__ == "__main__":
    main()

File: flink/KafkaCDCPrintJob.java
import org.apache.kafka.clients.consumer.ConsumerRecord;

import org.apache.flink.api.common.typeinfo.TypeInformation;
import org.apache.flink.api.common.eventtime.WatermarkStrategy;
import org.apache.flink.connector.kafka.source.KafkaSource;
import org.apache.flink.connector.kafka.source.enumerator.initializer.OffsetsInitializer;
import org.apache.flink.connector.kafka.source.reader.deserializer.KafkaRecordDeserializationSchema;
import org.apache.flink.streaming.api.datastream.DataStream;
import org.apache.flink.streaming.api.environment.StreamExecutionEnvironment;
import org.apache.flink.util.Collector;

import java.nio.charset.StandardCharsets;

public class KafkaCDCPrintJob {

    public static void main(String[] args) throws Exception {

        final StreamExecutionEnvironment env =
                StreamExecutionEnvironment.getExecutionEnvironment();

        KafkaRecordDeserializationSchema<String> deserializer =
                new KafkaRecordDeserializationSchema<String>() {

                    @Override
                    public void deserialize(
                            ConsumerRecord<byte[], byte[]> record,
                            Collector<String> out) {

                        // Ignore Kafka tombstone records
                        if (record.value() == null) {
                            return;
                        }

                        String value = new String(
                                record.value(),
                                StandardCharsets.UTF_8
                        );

                        out.collect(value);
                    }

                    @Override
                    public TypeInformation<String> getProducedType() {
                        return TypeInformation.of(String.class);
                    }
                };

        KafkaSource<String> source = KafkaSource.<String>builder()
                .setBootstrapServers("kafka:29092")
                .setTopics("sales.public.customers")
                .setGroupId("flink-cdc-print-test")
                .setStartingOffsets(OffsetsInitializer.earliest())
                .setDeserializer(deserializer)
                .build();

        DataStream<String> stream = env.fromSource(
                source,
                WatermarkStrategy.noWatermarks(),
                "Kafka CDC Source"
        );

        stream.print();

        env.execute("Kafka CDC Print Test");
    }
}
